shapely/coordinates.py

```diff
@@ -44,7 +90,55 @@ def transform(geometry, transformation, include_z=False):
     <POINT (1 1)>
     >>> transform(Point(0, 0, 0), lambda x: x + 1, include_z=True)
     <POINT Z (1 1 1)>
+
+    An identity function applicable to both types of input with single_vec=False
+
+    >>> g1 = LineString([(1, 2), (3, 4)])
+
+    >>> transform(g1, lambda x, y, z=None: tuple(filter(None, [x, y, z])), single_vec=False)
+    <LINESTRING (1 2, 3 4)>
+
+    example of another lambda expression:
+
+    >>> transform(g1, lambda x, y, z=None: (x+1.0, y+1.0), single_vec=False)
+    <LINESTRING (2 3, 4 5)>
+
+    Using pyproj >= 2.1, the following example will accurately project Shapely geometries
+    It transforms from EPSG:4326 (WGS84 log/lat) to EPSG:32618 (WGS84 UTM 18 North)
+    Note: always_xy kwarg is required as Shapely geometries only support X,Y coordinate ordering.
+
+    >>> try:
+    ...     import pyproj
+    ...     project = pyproj.Transformer.from_crs(4326, 32618, always_xy=True).transform
+    ...     p = transform(Point(-75, 50), project, single_vec=False)
+    ...     assert (round(p.x), round(p.y)) == (500000, 5538631)
+    ... except ImportError:
+    ...     pass
     """
+    if include_z == -1:
+        include_z = shapely.get_coordinate_dimension(geometry) == 3
+    if rebuild:
+        transform_wrapper = _transform2 if single_vec else _transform2_multi_vec
+        vectorize = not isinstance(geometry, Geometry)
+    else:
+        transform_wrapper = _transform if single_vec else _transform_multi_vec
+        vectorize = False
+        if not np.isscalar(include_z):
+            if np.all(include_z == include_z[0]):
+                include_z = include_z[0]
+            else:
+                vectorize = True
```

Comment:
Alternatively, instead of using np.frompyfunc (to effectively apply the function on each geometry separately, right?), another option is to do what we did in geopandas: https://github.com/geopandas/geopandas/blob/f17b5350b8fa8acb8926e78ee011f876c7a03ff4/geopandas/_vectorized.py#L1099-L1117

Subsetting the input array in 2D and 3D geometries, and then transforming each of those subsets in a single go still using the natively vectorized transform function (without doing the vectorize on the python level with np.frompyfunc)

(might be useful to compare both approaches for performance)

Reply:
Another option is also what we did in https://github.com/shapely/shapely/pull/1559, which is to infer `include_z` from the geometries (eg set it to True once there it at least one 3D geometry in the input), but then still use a single value for all geometries. 

For the `to_ragged_array` added in https://github.com/shapely/shapely/pull/1559 that is the only option (since the goal of that function is to return a single array). But of course here when it comes to transforming, you can easily have a transform function that cannot handle the NaNs very well (for example, in geopandas we ran into that with pyproj.Transformer.transform). So for the `shapely.transform` function, it probably _is_ useful to be able to have `include_z` per geometry.

Reply:
> Alternatively, instead of using np.frompyfunc (to effectively apply the function on each geometry separately, right?), another option is to do what we did in geopandas: https://github.com/geopandas/geopandas/blob/f17b5350b8fa8acb8926e78ee011f876c7a03ff4/geopandas/_vectorized.py#L1099-L1117
> 
> Subsetting the input array in 2D and 3D geometries, and then transforming each of those subsets in a single go still using the natively vectorized transform function (without doing the vectorize on the python level with np.frompyfunc)
> 
> (might be useful to compare both approaches for performance)

I thought that it might be faster to do as you suggested, but then I thought that it wouldn't work for the old non-vectorized `rebuild` version, and the approach I chose works for all the cases without extra code, and also it would would be more complicated to implement and review, so I figured I will leave this idea for future PR.

Reply:
> For the `to_ragged_array` added in #1559 that is the only option (since the goal of that function is to return a single array). But of course here when it comes to transforming, you can easily have a transform function that cannot handle the NaNs very well (for example, in geopandas we ran into that with pyproj.Transformer.transform). So for the `shapely.transform` function, it probably _is_ useful to be able to have `include_z` per geometry.

Yes, exactly I thought that deducting a single `include_z` value for `include_z=None` could yield inconsistent results (`transform(func, [point2d, point3d], include_z=None)` could potentially yield different results then  `[transform(func, point2d, include_z=None), transform(func,, point3d, include_z=None)]`, which doesn't sound right.